go/pkg/cas/upload_test.go

```diff
@@ -453,6 +470,185 @@ func TestStreaming(t *testing.T) {
 	}
 }
 
+func TestMissingDirs(t *testing.T) {
+	mustDigest := func(m proto.Message) *repb.Digest {
+		d, err := digest.NewFromMessage(m)
+		if err != nil {
+			t.Fatal(err)
+		}
+		return d.ToProto()
+	}
+
+	barDigest := digest.NewFromBlob([]byte("bar")).ToProto()
+	bazDigest := mustDigest(&repb.Directory{})
+
+	foo := &repb.Directory{
+		Files: []*repb.FileNode{{
+			Name:   "bar",
+			Digest: barDigest,
+		}},
+		Directories: []*repb.DirectoryNode{{
+			Name:   "baz",
+			Digest: bazDigest,
+		}},
+	}
+
+	root := &repb.Directory{
+		Directories: []*repb.DirectoryNode{{
+			Name:   "foo",
+			Digest: mustDigest(foo),
+		}},
+	}
+
+	testCases := []struct {
+		desc     string
+		tree     map[string]*digested
+		wantDirs []missingDir
+	}{
+		{
+			desc: "works",
+			tree: map[string]*digested{
+				"foo/bar": {
+					dirEntry: &repb.FileNode{
+						Name:   "bar",
+						Digest: barDigest,
+					},
+					digest: barDigest,
+				},
+				"foo/baz": {
+					dirEntry: &repb.DirectoryNode{
+						Name:   "baz",
+						Digest: bazDigest,
+					},
+					digest: bazDigest,
+				},
+			},
+			wantDirs: []missingDir{
+				{
+					relPath: ".",
+					item:    uploadItemFromDirMsg("/", root),
+					digested: &digested{
+						dirEntry: &repb.DirectoryNode{
+							Name:   "/",
+							Digest: mustDigest(root),
+						},
+						digest: mustDigest(root),
+					},
+				},
+				{
+					relPath: "foo",
+					item:    uploadItemFromDirMsg("/foo", foo),
+					digested: &digested{
+						dirEntry: &repb.DirectoryNode{
+							Name:   "foo",
+							Digest: mustDigest(foo),
+						},
+						digest: mustDigest(foo),
+					},
+				},
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		gotDirs := missingDirs(tc.tree, "/")
+		sort.Slice(gotDirs, func(i, j int) bool {
+			return gotDirs[i].relPath < gotDirs[j].relPath
+		})
+
+		compareDirs := func(x, y missingDir) bool {
+			return proto.Equal(x.digested.digest, y.digested.digest) &&
+				proto.Equal(x.digested.dirEntry, y.digested.dirEntry) &&
+				compareUploadItems(x.item, y.item)
```

Comment:
Is it possible to pass only this function to `cmp.Comparer` with `protocmp.Transform` to `cmp.Diff`?

Reply:
I can't figure out how to do that. How would you do that?

Reply:
no longer applicable